### PR TITLE
Create streaming interface for client streaming.

### PR DESCRIPTION
**Pull request by wwitzel3**
**What this PR does / why we need it**:
Extracts the current websocket streaming in to an interface that can be replaced.
This work was done by @mstergianis in #1905 this PR is just addressing the merge conflict.

### CODE REVIEW

File: pkg/dash/dash.go [hunk 1]
```diff
@@ -215,15 +216,23 @@ func WithClusterClient(client cluster.ClientInterface) RunnerOption {
 	}
 }
 
+func WithStreamingClientFactory(factory api.StreamingClientFactory) RunnerOption {
```
Comment (jamieklassen) [inline pkg/dash/dash.go:219]: the package `api` in this context is an `internal` package -- for library consumers to use this option the `StreamingClientFactory` type will need to be moved into a non-internal package please!

**Review by jamieklassen — CHANGES_REQUESTED** [review]
requesting one change so that library consumers can effectively use this handy feature!

**Comment by mstergianis** [comment]
@wwitzel3 happy to resolve the above two requests. I'll send you a patch

**Comment by mstergianis** [comment]
@wwitzel3 in going through the exercise of resolving the requests; I realized that `StreamError` should also be part of pkg as a custom implementation should have the option to return a fatal stream error. Wanted to get your thoughts on moving this much implementation out of `internal` before I spend more time on it.

**Comment by GuessWhoSamFoo** [comment]
Making `StreamError` public is fine - realistically this is only going to be consumed internally by folks at VMware.

Eventually, `internal/errors` should be public as well and can have a similar API to `pkg/store` so third party usage or tools can add warnings/errors that will be visible by UI.

File: internal/api/streaming_connection_manager.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+type clientMeta struct {
+	cancelFunc context.CancelFunc
+	client     StreamingClient
+}
+
+type StreamingClientFactory interface {
+	NewConnection(uuid.UUID, http.ResponseWriter, *http.Request, *StreamingConnectionManager, config.Dash) (StreamingClient, context.CancelFunc, error)
+	NewTemporaryConnection(uuid.UUID, http.ResponseWriter, *http.Request, *StreamingConnectionManager) (StreamingClient, context.CancelFunc, error)
```
Comment (jamieklassen) [inline internal/api/streaming_connection_manager.go:40]:
```suggestion
	NewConnection(http.ResponseWriter, *http.Request, *StreamingConnectionManager, config.Dash) (StreamingClient, context.CancelFunc, error)
	NewTemporaryConnection(http.ResponseWriter, *http.Request, *StreamingConnectionManager) (StreamingClient, context.CancelFunc, error)
```

File: internal/api/streaming_connection_manager.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+// ClientFromRequest creates a websocket client from a http request.
+func (m *StreamingConnectionManager) ClientFromRequest(dashConfig config.Dash, w http.ResponseWriter, r *http.Request) (StreamingClient, error) {
+	clientID, err := uuid.NewUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	client, cancel, err := m.clientFactory.NewConnection(clientID, w, r, m, dashConfig)
```
Comment (jamieklassen) [inline internal/api/streaming_connection_manager.go:127]:
```suggestion
	client, cancel, err := m.clientFactory.NewConnection(w, r, m, dashConfig)
```

File: internal/api/streaming_connection_manager.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	return client, nil
+}
+
+func (m *StreamingConnectionManager) TemporaryClientFromLoadingRequest(w http.ResponseWriter, r *http.Request) (StreamingClient, error) {
+	clientID, err := uuid.NewUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	client, cancel, err := m.clientFactory.NewTemporaryConnection(clientID, w, r, m)
```
Comment (jamieklassen) [inline internal/api/streaming_connection_manager.go:148]:
```suggestion
	client, cancel, err := m.clientFactory.NewTemporaryConnection(w, r, m)
```

File: internal/api/websocket_connection_factory.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			},
+		},
+	}
+}
+
+var _ StreamingClientFactory = (*WebsocketConnectionFactory)(nil)
+
+func (wcf *WebsocketConnectionFactory) NewConnection(
+	clientID uuid.UUID, w http.ResponseWriter, r *http.Request, m *StreamingConnectionManager, dashConfig config.Dash,
+) (StreamingClient, context.CancelFunc, error) {
```
Comment (jamieklassen) [inline internal/api/websocket_connection_factory.go:39]:
```suggestion
func (wcf *WebsocketConnectionFactory) NewConnection(
	w http.ResponseWriter, r *http.Request, m *StreamingConnectionManager, dashConfig config.Dash,
) (StreamingClient, context.CancelFunc, error) {
 	clientID, err := uuid.NewUUID()
	if err != nil {
		return nil, nil, err
	}
```

File: internal/api/websocket_connection_factory.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+	go client.readPump()
+	go client.writePump()
+
+	return client, cancel, nil
+}
+
+func (wcf *WebsocketConnectionFactory) NewTemporaryConnection(
+	clientID uuid.UUID, w http.ResponseWriter, r *http.Request, m *StreamingConnectionManager,
+) (StreamingClient, context.CancelFunc, error) {
```
Comment (jamieklassen) [inline internal/api/websocket_connection_factory.go:56]:
```suggestion
func (wcf *WebsocketConnectionFactory) NewTemporaryConnection(
	w http.ResponseWriter, r *http.Request, m *StreamingConnectionManager,
) (StreamingClient, context.CancelFunc, error) {
 	clientID, err := uuid.NewUUID()
	if err != nil {
		return nil, nil, err
	}
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
I'd like to suggest that we move the generation of IDs into the connection factory itself -- it reduces the surface area of the connection factory interface and allows custom implementations (like the one I'd like to write) to manage their own IDs.

**Comment by mstergianis** [comment]
Apologies, I made the changes requested above and forgot to update this thread.

I've made a PR against @wwitzel3's [branch](https://github.com/wwitzel3/octant/pull/240). The diffs ended up being quite large so I thought that the easiest way to make the change.

**Comment by wwitzel3** [comment]
replaced by #2581

**Pull request closed without merging**